manifests/server/config/memory.pp

```diff
@@ -0,0 +1,138 @@
+        clickhouse::error { 'This host doesn\'t have sufficient amount of RAM, OOMs are possible':
+        }
+
+        $max_memory_usage_for_all_queries_auto = 0
+        $max_memory_usage_auto = 0
+        $uncompressed_cache_size_auto = 8 << 30 # Default value from CH config
+        $max_bytes_before_external_group_by = 0
+        $max_bytes_before_external_sort = 0
+    } else {
+        # This block is trying setup memory consumption automatically if *_b parameters are unset
+        $max_memory_usage_for_all_queries_auto = (!!$max_memory_usage_for_all_queries) ? {
+            true    => $max_memory_usage_for_all_queries,
+            default => (!!$uncompressed_cache_size) ? {
+                true    => $efficient_memory - $uncompressed_cache_size,
+                default => $efficient_memory * 5 / 6,
+            }
+        }
+
+        $max_memory_usage_auto = (!!$max_memory_usage) ? {
+            true    => $max_memory_usage,
+            default => $max_memory_usage_for_all_queries_auto,
+        }
+
+        $uncompressed_cache_size_auto = (!!$uncompressed_cache_size) ? {
+            true    => $uncompressed_cache_size,
+            default => $efficient_memory - $max_memory_usage_for_all_queries_auto,
+        }
+
+        $max_bytes_before_external_group_by = $external_group_by ? {
+            true    => $max_memory_usage_auto / 2,
+            default => 0,
+        }
+
+        $max_bytes_before_external_sort = $external_sort ? {
+            true    => $max_memory_usage_auto / 2,
+            default => 0,
+        }
+
+        with($max_memory_usage_for_all_queries_auto + $uncompressed_cache_size_auto) |$used| {
+            if $used != $efficient_memory and $memory_check {
+                notify { 'Memory':
+                    message  => "The memory configuration for ClickHouse is inefficient. Avaliable: ${efficient_memory}; used: ${used}",
+                    loglevel => 'warning',
+                }
+            }
+        }
+    }
+
+    $default_memory = {
+        'profiles' => {
+            'default' => {
+                'use_uncompressed_cache'                   => [$use_uncompressed_cache],
+                'max_memory_usage'                         => [$max_memory_usage_auto],
+                'max_memory_usage_for_all_queries'         => [$max_memory_usage_for_all_queries_auto],
+                'max_bytes_before_external_group_by'       => [$max_bytes_before_external_group_by],
+                'max_bytes_before_external_sort'           => [$max_bytes_before_external_sort],
+            },
+        },
+    }
+
+    $memory_settings = {
+        'uncompressed_cache_size' => [$uncompressed_cache_size_auto], # cache for short queries
+        'mark_cache_size'         => [$mark_cache_size], # cache of "index"
+    }
+
+    clickhouse::server::config { 'default-memory':
+        section        => 'users',
+        data           => $default_memory,
+        service_notify => true,
```

Comment:
why are you enforcing this? and below. actually the user of this module should be able to set this somewhere globally (maybe in server class) and then it should only restart if the user wants it

Reply:
Good thing. Yes, I'll add some attribute to clickhouse::server, that will control `restarts when necessary to apply changes`